overlord/configstate/configcore/certs.go

```diff
@@ -32,6 +33,221 @@ import (
+}
+
+// This function validates the name of the certificate. We only allow certificates
+// with the correct suffix ".crt" except for the ca-certificates.crt,
+// or if the name is in the blockedCerts lists.
+func isBlocked(cert certificate, blockedCerts []string) bool {
+	// Special case for ca-certificates.crt
+	if cert.Name == "ca-certificates.crt" {
+		return true
+	}
+
+	// Check that the real underlying filepath to the certificate ends with .crt
+	if !strings.HasSuffix(cert.RealPath, ".crt") {
+		return true
+	}
+
+	// See if the name appears in the blockedCerts list
+	for _, blockedCert := range blockedCerts {
+		if cert.Name == blockedCert {
+			return true
+		}
+	}
+	return false
+}
+
+// TODO Should we support recursive directories inside the certs dir?
+func getCertObjects(basePath string) ([]certificate, error) {
+	certFiles, err := ioutil.ReadDir(basePath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read base certs: %v", err)
+	}
+
+	certsObjects := []certificate{}
+	for _, cert := range certFiles {
+		if cert.IsDir() {
+			continue
+		}
+
+		// So what are we trying to achieve here? When provided with certificate directories
+		// they may or may not be symbolic links to the actual certificate file. To handle this
+		// correctly we evaluate the symbolic links and get the real path of the certificate, because
+		// we are not going to 'copy' or link to the symbolic link, instead we will recreate the symbolic
+		// link to the real path of the certificate.
+		certRealPath := filepath.Join(basePath, cert.Name())
+		if cert.Mode()&os.ModeSymlink != 0 {
+			resolvedPath, err := filepath.EvalSymlinks(certRealPath)
+			if err == nil {
+				certRealPath = resolvedPath
+			}
+		}
+
+		// If the file is not a symbolic link then Path and RealPath will be identical.
+		certObject := certificate{
+			Name:     cert.Name(),
+			Path:     filepath.Join(basePath, cert.Name()),
+			RealPath: certRealPath,
+		}
+		certsObjects = append(certsObjects, certObject)
+	}
+	return certsObjects, nil
+}
+
+// Filters out the certificates that are not allowed to be merged.
+func filterCerts(certs []certificate, blockedCerts []string) ([]certificate, error) {
+	filteredCerts := []certificate{}
+	for _, cert := range certs {
+		if isBlocked(cert, blockedCerts) {
+			continue
+		}
+		filteredCerts = append(filteredCerts, cert)
+	}
+	return filteredCerts, nil
+}
+
+func filterCertsInDirectory(dirPath string, blockedCerts []string) error {
+	existingCerts, err := getCertObjects(dirPath)
+	if err != nil {
+		return err
+	}
+
+	// Remove the certs that are marked blocked.
+	for _, cert := range existingCerts {
+		if isBlocked(cert, blockedCerts) {
+			err := os.Remove(cert.Path)
+			if err != nil {
+				return fmt.Errorf("cannot remove blocked certificate: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+// Populates symbolic links in the output directory, to each certificate
+// provided.
+func installCerts(outputPath string, certs []certificate) error {
```

Comment:
Why do we prefer symbolic links to copying the files? Doesn't that increase the chances of something modifying the real file while not updating `ca-certificates.crt`, for instance?

Reply:
I suppose ca-certificate package prefers symlinks in `/etc` to simplify update merging and avoid users to modify the files. I wonder why we actually need those files. Is there any ssl library that reads all the files in a directory instead of the `ca-certificates.crt` bundle?

Reply:
This is what is done by update-ca-certificates, it creates symlinks - so I just made sure to keep that functionality. You do raise an interesting point about what happens here. I would love some other input as well here as I have to admit I don't know what is the right thing to do here.

@mvo5 Can you help here maybe?

Reply:
@valentindavid good point about updates, I'm thinking now that does makes sense in the deb world, but maybe not that much for UC, where probably certs will be added by `snap get/set` actions.

Reply:
if we keep symlinking we'll need to apparmor rules that give access to the source directories and not only the merged directory, maybe that's fine but something to keep in minde